pytest/tests/sanity/repro_2916.py

```diff
@@ -0,0 +1,107 @@
+# track).
+# 
+# We then kill both nodes, and restart the first node, and do the same
+# requests. We expect it to resond the same way. Before 2916 is fixed, it
+# fails to respond to the request it was previously responding to due to
+# incorrect reconstruction of the receipts.
+
+import asyncio, sys, time
+import socket, base58
+import nacl.signing, hashlib
+
+sys.path.append('lib')
+
+from cluster import start_cluster
+from peer import *
+from utils import obj_to_string
+
+from messages.tx import *
+from messages.block import *
+from messages.crypto import *
+from messages.network import *
+
+async def main():
+    # start a cluster with two shards
+    nodes = start_cluster(2, 0, 2, None, [], {})
+
+    started = time.time()
+
+    while True:
+        if time.time() - started > 10:
+            assert False, "Giving up waiting for two blocks"
+
+        status = nodes[0].get_status()
+        hash_ = status['sync_info']['latest_block_hash']
+        height = status['sync_info']['latest_block_height']
+
+        if height > 2:
+            block = nodes[0].get_block(hash_)
+            chunk_hashes = [base58.b58decode(x['chunk_hash']) for x in block['result']['chunks']]
+
+            assert len(chunk_hashes) == 2
+            assert all([len(x) == 32 for x in chunk_hashes])
+
+            break
+
+    my_key_pair_nacl = nacl.signing.SigningKey.generate()
+    received_responses = [None, None]
+
+# step = 0: before the node is killed
+# step = 1: after the node is killed
+    for step in range(2):
+
+        conn0 = await connect(nodes[0].addr())
+        await run_handshake(conn0, nodes[0].node_key.pk, my_key_pair_nacl)
+        for shard_ord, chunk_hash in enumerate(chunk_hashes):
+
+            request = PartialEncodedChunkRequestMsg()
+            request.chunk_hash = chunk_hash
+            request.part_ords = []
+            request.tracking_shards = [0, 1]
+
+            routed_msg_body = RoutedMessageBody()
+            routed_msg_body.enum = 'PartialEncodedChunkRequest'
+            routed_msg_body.PartialEncodedChunkRequest = request
+
+            peer_message = create_and_sign_routed_peer_message(routed_msg_body, nodes[0], my_key_pair_nacl)
+
+            await conn0.send(peer_message)
+
+            received_response = False
+            # make several attempts, in case the node sends us a block or some other message before sending the response
+            for i in range(5):
+                response = await conn0.recv()
+                if response is None: # None means timeout
+                    break
+                elif response.enum == 'Routed' and response.Routed.body.enum == 'PartialEncodedChunkResponse':
+                    print("Received response for shard %s" % shard_ord)
+                    received_response = True
+                    break
+                else:
+                    print(response.enum)
+            else:
+                print("Didn't receive response for shard %s" % shard_ord)
```

Comment:
as mentioned by @bowenwang1996 this might not be very reliable. Instead we can pass a filter to `recv` such that it waits until the expected message arrives. As per suggestion above, we can use `expect_fn` like the following snippet:

```suggestion
            try:
                expected_fn = lambda response : response.enum == 'Routed' and response.Routed.body.enum == 'PartialEncodedChunkResponse' 
                _response = await asyncio.wait_for(conn0.recv(expected_fn), 5) 
                recieved_response = True
                break
            except concurrent.futures.TimeoutError:
                print("Didn't receive response for shard %s" % shard_ord)
```

This way we only have a timeout in one place, and to wait exactly for the message we are expecting, this will work even if we receive in between several not relevant messages, as long as the expected message arrives before the timeout.